File: packages/core/src/policy.ts
import type {
  ArgConstraint,
  AuthorizationDetail,
  Budget,
  Policy,
  PolicyRule,
} from "./types.js";

export interface ToolCallContext {
  upstream: string;
  tool: string;
  args: Record<string, unknown>;
  costUnits: number;
}

export type DecisionEffect = "allow" | "deny" | "require_approval";
export type DecisionSource = "token_bounds" | "rule" | "constraint" | "budget" | "default";

export interface Decision {
  effect: DecisionEffect;
  source: DecisionSource;
  reason: string;
  ruleId?: string;
}

/**
 * Full gate decision: the token's authorization_details bound what is even
 * *reachable*; policy rules then decide among reachable calls; default is deny.
 * Budget is checked separately (it is stateful) via `checkBudget`.
 */
export function decide(
  policy: Policy,
  call: ToolCallContext,
  tokenAuthorization: AuthorizationDetail[],
): Decision {
  if (!isWithinAuthorization(call, tokenAuthorization)) {
    return {
      effect: "deny",
      source: "token_bounds",
      reason: `tool ${call.upstream}.${call.tool} is outside the token's authorization_details`,
    };
  }
  return evaluatePolicy(policy, call);
}

export function isWithinAuthorization(
  call: Pick<ToolCallContext, "upstream" | "tool">,
  details: AuthorizationDetail[],
): boolean {
  return details.some(
    (d) => d.upstream === call.upstream && (d.tools.includes("*") || d.tools.includes(call.tool)),
  );
}

export function evaluatePolicy(policy: Policy, call: ToolCallContext): Decision {
  for (const [index, rule] of policy.rules.entries()) {
    if (!ruleMatches(rule, call)) continue;
    const ruleId = rule.id ?? `${policy.id}#${index}`;

    const maxCost = rule.constraints?.maxCostUnits;
    if (rule.effect === "allow" && maxCost !== undefined && call.costUnits > maxCost) {
      return {
        effect: "deny",
        source: "constraint",
        ruleId,
        reason: `call cost ${call.costUnits} exceeds rule maxCostUnits ${maxCost}`,
      };
    }

    return {
      effect: rule.effect,
      source: "rule",
      ruleId,
      reason: rule.description ?? `matched ${rule.effect} rule ${ruleId}`,
    };
  }
  return { effect: "deny", source: "default", reason: "no policy rule matched (default deny)" };
}

function ruleMatches(rule: PolicyRule, call: ToolCallContext): boolean {
  if (rule.match.upstream !== undefined && !globMatch(rule.match.upstream, call.upstream)) {
    return false;
  }
  if (rule.match.tool !== undefined && !globMatch(rule.match.tool, call.tool)) {
    return false;
  }
  if (rule.match.where !== undefined) {
    return rule.match.where.every((c) => constraintHolds(c, call.args));
  }
  return true;
}

/** Minimal glob: "*" matches any run of characters; everything else is literal. */
export function globMatch(pattern: string, value: string): boolean {
  if (pattern === "*") return true;
  const escaped = pattern.split("*").map(escapeRegExp).join(".*");
  return new RegExp(`^${escaped}$`).test(value);
}

function escapeRegExp(s: string): string {
  return s.replace(/[.*+?^${}()|[\]\\]/g, "\\$&");
}

export function constraintHolds(constraint: ArgConstraint, args: Record<string, unknown>): boolean {
  const actual = getPath(args, constraint.path);
  const expected = constraint.value;
  switch (constraint.op) {
    case "eq":
      return actual === expected;
    case "neq":
      return actual !== expected;
    case "gt":
      return typeof actual === "number" && typeof expected === "number" && actual > expected;
    case "gte":
      return typeof actual === "number" && typeof expected === "number" && actual >= expected;
    case "lt":
      return typeof actual === "number" && typeof expected === "number" && actual < expected;
    case "lte":
      return typeof actual === "number" && typeof expected === "number" && actual <= expected;
    case "in":
      return Array.isArray(expected) && expected.some((v) => v === actual);
    case "contains":
      if (typeof actual === "string" && typeof expected === "string") {
        return actual.includes(expected);
      }
      return Array.isArray(actual) && actual.some((v) => v === expected);
    case "startsWith":
      return (
        typeof actual === "string" && typeof expected === "string" && actual.startsWith(expected)
      );
    case "matches":
      return (
        typeof actual === "string" &&
        typeof expected === "string" &&
        new RegExp(expected).test(actual)
      );
  }
}

export function getPath(obj: Record<string, unknown>, path: string): unknown {
  let current: unknown = obj;
  for (const segment of path.split(".")) {
    if (current === null || typeof current !== "object") return undefined;
    current = (current as Record<string, unknown>)[segment];
  }
  return current;
}

export interface BudgetCheck {
  ok: boolean;
  remainingUnits: number;
}

export function checkBudget(budget: Budget, costUnits: number): BudgetCheck {
  const remaining = budget.maxUnits - budget.spentUnits;
  return { ok: costUnits <= remaining, remainingUnits: remaining };
}

File: src/toolgate/core/policy.py
import re
from dataclasses import dataclass
from typing import Any, Literal

from .types import (
    ArgConstraint,
    AuthorizationDetail,
    Budget,
    DecisionEffect,
    DecisionSource,
    Policy,
    PolicyRule,
)


@dataclass(frozen=True)
class ToolCallContext:
    upstream: str
    tool: str
    args: dict[str, Any]
    cost_units: int


@dataclass(frozen=True)
class Decision:
    effect: DecisionEffect
    source: DecisionSource
    reason: str
    rule_id: str | None = None


def decide(
    policy: Policy,
    call: ToolCallContext,
    token_authorization: list[AuthorizationDetail],
) -> Decision:
    """Full gate decision: the token's authorization_details bound what is even
    *reachable*; policy rules then decide among reachable calls; default is
    deny. Budget is checked separately (it is stateful) via `check_budget`."""
    if not is_within_authorization(call, token_authorization):
        return Decision(
            effect="deny",
            source="token_bounds",
            reason=(
                f"tool {call.upstream}.{call.tool} is outside the token's authorization_details"
            ),
        )
    return evaluate_policy(policy, call)


def is_within_authorization(
    call: ToolCallContext, details: list[AuthorizationDetail]
) -> bool:
    return any(
        d.upstream == call.upstream and ("*" in d.tools or call.tool in d.tools)
        for d in details
    )


def evaluate_policy(policy: Policy, call: ToolCallContext) -> Decision:
    for index, rule in enumerate(policy.rules):
        if not _rule_matches(rule, call):
            continue
        rule_id = rule.id or f"{policy.id}#{index}"

        max_cost = rule.constraints.maxCostUnits if rule.constraints else None
        if rule.effect == "allow" and max_cost is not None and call.cost_units > max_cost:
            return Decision(
                effect="deny",
                source="constraint",
                rule_id=rule_id,
                reason=f"call cost {call.cost_units} exceeds rule maxCostUnits {max_cost}",
            )

        return Decision(
            effect=rule.effect,
            source="rule",
            rule_id=rule_id,
            reason=rule.description or f"matched {rule.effect} rule {rule_id}",
        )
    return Decision(effect="deny", source="default", reason="no policy rule matched (default deny)")


def _rule_matches(rule: PolicyRule, call: ToolCallContext) -> bool:
    if rule.match.upstream is not None and not glob_match(rule.match.upstream, call.upstream):
        return False
    if rule.match.tool is not None and not glob_match(rule.match.tool, call.tool):
        return False
    if rule.match.where is not None:
        return all(constraint_holds(c, call.args) for c in rule.match.where)
    return True


def glob_match(pattern: str, value: str) -> bool:
    """Minimal glob: "*" matches any run of characters; everything else is literal."""
    if pattern == "*":
        return True
    escaped = ".*".join(re.escape(part) for part in pattern.split("*"))
    return re.fullmatch(escaped, value) is not None


def constraint_holds(constraint: ArgConstraint, args: dict[str, Any]) -> bool:
    actual = get_path(args, constraint.path)
    expected = constraint.value
    op = constraint.op
    if op == "eq":
        return actual == expected
    if op == "neq":
        return actual != expected
    if op in ("gt", "gte", "lt", "lte"):
        if not _both_numbers(actual, expected):
            return False
        if op == "gt":
            return actual > expected
        if op == "gte":
            return actual >= expected
        if op == "lt":
            return actual < expected
        return actual <= expected
    if op == "in":
        return isinstance(expected, list) and actual in expected
    if op == "contains":
        if isinstance(actual, str) and isinstance(expected, str):
            return expected in actual
        return isinstance(actual, list) and expected in actual
    if op == "startsWith":
        return (
            isinstance(actual, str) and isinstance(expected, str) and actual.startswith(expected)
        )
    if op == "matches":
        return (
            isinstance(actual, str)
            and isinstance(expected, str)
            and re.search(expected, actual) is not None
        )
    return False


def _both_numbers(a: Any, b: Any) -> bool:
    # bool is an int subclass in Python; a numeric comparison against True/False
    # is never an intended policy, so exclude it explicitly.
    return (
        isinstance(a, (int, float))
        and not isinstance(a, bool)
        and isinstance(b, (int, float))
        and not isinstance(b, bool)
    )


def get_path(obj: dict[str, Any], path: str) -> Any:
    current: Any = obj
    for segment in path.split("."):
        if not isinstance(current, dict):
            return None
        current = current.get(segment)
    return current


@dataclass(frozen=True)
class BudgetCheck:
    ok: bool
    remaining_units: int


def check_budget(budget: Budget, cost_units: int) -> BudgetCheck:
    remaining = budget.maxUnits - budget.spentUnits
    return BudgetCheck(ok=cost_units <= remaining, remaining_units=remaining)


EffectLiteral = Literal["allow", "deny", "require_approval"]
